include/nuttx/serial/tioctl.h

```diff
@@ -190,6 +191,7 @@
 #  define SER_SINGLEWIRE_PULL_DISABLE (0 << SER_SINGLEWIRE_PULL_SHIFT) /* Float RX/TX Line */
 #  define SER_SINGLEWIRE_PULLUP       (1 << SER_SINGLEWIRE_PULL_SHIFT) /* Enable Pull up the RX/TX Line */
 #  define SER_SINGLEWIRE_PULLDOWN     (2 << SER_SINGLEWIRE_PULL_SHIFT) /* Enable Pull down the RX/TX Line */
+#  define SER_SINGLEWIRE_PUSHPULL     (1 << 3)                         /* Use PUSH/PULL not Open Drain with Single wire */
```

Comment:
should we extend SER_SINGLEWIRE_PULL_MASK to 7 << SER_SINGLEWIRE_PULL_SHIFT, and define SER_SINGLEWIRE_PUSHPULL to 3 << SER_SINGLEWIRE_PULL_SHIFT? otherwise, how the driver handle both SER_SINGLEWIRE_PUSHPULL and SER_SINGLEWIRE_PULLDOWN set at the same time?

Reply:
@xiaoxiang781216 not sure I follow...

enable:0000 0001
pull mask   : 0000 0110
       pull up: 0000 0010
       pull dn: 0000 0100
pushpull    : 0000  1000

pp w/   pull dn: 0000 1101

Reply:
Can a pin be in PULLUP and PULLDOWN at the same time? I don't think so:
```
gpio_val |= (arg & SER_SINGLEWIRE_PULL_MASK) == SER_SINGLEWIRE_PULLUP ? GPIO_PULLUP : GPIO_FLOAT;
            gpio_val |= (arg & SER_SINGLEWIRE_PULL_MASK) == SER_SINGLEWIRE_PULLDOWN ? GPIO_PULLDOWN : GPIO_FLOAT;
```
The code check is done by == and &, also indicate that we should use PULLUP/PULLDOWN as value not flag.

Reply:
@xiaoxiang781216 I am sorry I am not following your objection

> Can a pin be in PULLUP and PULLDOWN at the same time? I don't think so 

No.  

But it can be a Push Pull or Open drain with PU or PD 

Calling code is 
```
if (ioctl(uart, TIOCSSINGLEWIRE, single_wire ? (SER_SINGLEWIRE_ENABLED  |  SER_SINGLEWIRE_PUSHPULL | SER_SINGLEWIRE_PULLDOWN) : 0) < 0) 
```
if some did it wrong like
```
if (ioctl(uart, TIOCSSINGLEWIRE, single_wire ? (SER_SINGLEWIRE_ENABLED  |  SER_SINGLEWIRE_PUSHPULL | SER_SINGLEWIRE_PULLDOWN | SER_SINGLEWIRE_PULLDOWN) : 0) < 0) 
```
They get pushpull float.

The arg is treated as a bit mask. with bit fields 1,2,1 is that not clear?

Reply:
PUSHPULL normally mean drive the line with high/low voltage, I can't understand why it need ored with PULLUP/PULLDOWN? What's mean if the caller pass PUSHPULL | PULLUP?

Reply:
This is the typical IO pad structure. 

![image](https://user-images.githubusercontent.com/1945821/89402895-61c20880-d6cc-11ea-867c-8bb3bda08d7b.png)

In single wire the TX pin is both an input and an output. 
![image](https://user-images.githubusercontent.com/1945821/89403939-fbd68080-d6cd-11ea-932b-f3640c5f3106.png)

The Second highlighted portion is what is referenced in Summary above by ". Therefor it is impossible to meet proper signaling level thresholds by following the reference manual, and configuring the TX for Open drain with a pull up."

In the case of the inverted protocol  it pulls down the pin while it is an input and helps with the signal margins. 

This was explained in Summary above "Adding the pull down increases margins.", the low is .1V lower with the added pull down.  This increase that signal margins. 

> What's mean if the caller pass PUSHPULL | PULLUP?

For single wire it means that while in input mode the pin would be pulled high if not driven.

I hope this is clear and you can get this merged. If not possibly @liguiding can help clarify it for you.

Reply:
Ok.